metomi/isodatetime/data.py

```diff
@@ -612,58 +738,90 @@ def __str__(self):
+        TimeZone. Some operations and comparisons may fail when this is True.
+    _is_empty_instance (bool): If True, do not set any properties yet. These
+        should be set as part of a copy operation.
     """
 
-    __slots__ = ['unknown'] + Duration.__slots__
+    __slots__ = [*Duration.__slots__, "_unknown"]
 
-    def __init__(self, hours=0, minutes=0, unknown=False):
-        self.unknown = unknown
-        super(TimeZone, self).__init__(hours=hours, minutes=minutes)
+    to_weeks = property(doc='Unavailable/not inherited')
 
-    def copy(self):
-        """Return an unlinked copy of this instance."""
-        return TimeZone(hours=self.hours, minutes=self.minutes,
-                        unknown=self.unknown)
+    def __init__(self, hours=0, minutes=0, unknown=False,
+                 _is_empty_instance=False):
+        if _is_empty_instance:
+            return
+        if hours is None:
+            hours = 0
+        else:
+            hours = _int_caster(hours, "TimeZone hours")
+            _bounds_checker(hours, "TimeZone hours",
+                            min_val=-99, max_val=99)
+        if minutes is None:
+            minutes = 0
+        else:
+            minutes = _int_caster(minutes, "TimeZone minutes")
+            min_minutes = 1 - CALENDAR.MINUTES_IN_HOUR
+            max_minutes = CALENDAR.MINUTES_IN_HOUR - 1
+            if hours > 0:
+                min_minutes = 0
+            elif hours < 0:
+                max_minutes = 0
+            _bounds_checker(minutes, "TimeZone minutes",
+                            min_val=min_minutes, max_val=max_minutes)
+        self._unknown = unknown
+        self._hours = hours
+        self._minutes = minutes
+        for attr in ["_years", "_months", "_days", "_seconds"]:
+            setattr(self, attr, 0)
+        self._weeks = None
+
+    @property
+    def unknown(self): return self._unknown
+
+    def __hash__(self) -> int:
+        # TODO: Do we have to worry about the possibility of a hash collision
+        # between instances of two different classes?
+        return hash((self._unknown, self._hours, self._minutes))
 
     def __str__(self):
-        if self.unknown:
+        if self._unknown:
             return ""
-        if self.hours == 0 and self.minutes == 0:
+        if self._hours == 0 and self._minutes == 0:
             return "Z"
         else:
             time_string = "+%02d:%02d"
-            if self.hours < 0 or (self.hours == 0 and self.minutes < 0):
+            if self._hours < 0 or (self._hours == 0 and self._minutes < 0):
                 time_string = "-%02d:%02d"
-            return time_string % (abs(self.hours), abs(self.minutes))
+            return time_string % (abs(self._hours), abs(self._minutes))
 
     def __repr__(self):
-        return "<isodatetime.data.TimeZone:" + repr(str(self)) + ">"
+        return "<isodatetime.data.TimeZone: {0}>".format(repr(str(self)))
```

Comment:
See previous comments about using values from `__self__` instead of hard-coding? (Although thinking again now, this PR is about making it hashable, so I think we can move it for later)